sharktank/sharktank/evaluate/perplexity_vmfb.py

```diff
@@ -0,0 +1,453 @@
+        )
+        vmfb_path = export_artifacts.get_artifacts()
+        return vmfb_path
+
+    @timeit
+    def load_model(self, weight_path, tokenizer, vmfb_path):
+
+        config = LlamaModelConfig(
+            hp=configs.LlamaHParams.from_gguf_props(weight_path.properties),
+            block_seq_stride=16,
+            kv_cache_type=self.kv_cache_type,
+            device=self.torch_device,
+            activation_dtype=self.activation_dtype,
+            attention_dtype=self.attention_dtype,
+            tensor_parallelism_size=self.tensor_parallelism_size,
+        )
+
+        if config.tensor_parallelism_size > 1:
+            weight_path.root_theta = shard_theta(weight_path.root_theta, config)
+
+        theta = weight_path.root_theta
+
+        if config.hp.expert_count:
+            if config.hp.model_arch == "grok":
+                model = PagedGrokModelV1(theta, config)
+            else:
+                model = PagedMixtralModelV1(theta, config)
+        else:
+            model = PagedLlamaModelV1(theta, config)
+
+        self.generator = TorchGenerator(model, tokenizer)
+
+        self.runner = vmfbRunner(
+            device=self.iree_device,
+            vmfb_path=vmfb_path,
+            external_weight_path=self.weight_path_str,
+        )
+
+    @timeit
+    def get_prompts(self):
+        test_prompts = load_dataset("wikitext", "wikitext-2-raw-v1", split="test")[
+            "text"
+        ]
+        num_test_prompts = 219
+
+        random.seed(0)
+        test_prompts = random.sample(test_prompts, num_test_prompts)
+
+        # Ignore prompts that are: empty, less than 20 tokens or a title.
+        test_prompts = [
+            s.replace("\n", "").rstrip()
+            for s in test_prompts
+            if s != "" and len(s.split()) >= 20 and s.count("=") < 2
+        ]
+
+        self.bs = len(test_prompts)
+
+        return test_prompts
+
+    def prefill_vmfb(self, token_batch, i):
+
+        logger.debug(f"Prefill:")
+
+        logger.debug("Input:")
+        logger.debug(f"{self.generator.tokenizer.decode(token_batch)}")
+
+        token_batch, seq_lens_batch = self.generator.tokenizer.pad_tokens(
+            token_ids=token_batch.tolist(),
+            pad_to_multiple_of=self.generator.model.cache.pad_sequence_stride,
+        )
+
+        logger.debug(f"{token_batch}")
+
+        token_batch = torch.tensor(token_batch, device=self.torch_device)
+        self.seq_lens_batch = torch.tensor(seq_lens_batch, device=self.torch_device)
+
+        self.batch = self.generator.begin_eval_batch(
+            token_batch=token_batch,
+            seq_lens_batch=self.seq_lens_batch,
+            bs=self.bs,
+        )
+
+        seq_block_ids = self.batch.pad_block_ids()
+        prefill_logits = self.runner.ctx.modules.module[f"prefill_bs{self.bs}"](
+            token_batch,
+            self.seq_lens_batch,
+            seq_block_ids,
+            self.batch.cache_state[0].to(torch.float16),
```

Comment:
@archana-ramalingam, I think this may make a copy.
First the .to(...) may make a copy.
Then before the making the actual call to the IREE module function it will make a copy to the device if we are not targeting the CPU.

Reply:
You would need to first explicitly make a `iree.runtime.DeviceArray` and then copy back to the cache state after the call.

I have also added some IREE related functionality [here](https://github.com/nod-ai/SHARK-Platform/blob/main/sharktank/sharktank/utils/iree.py). We should unify this aspect to reduce code duplication.